js/services/Vocabulary.js

```diff
@@ -12,39 +12,47 @@ define(function (require, exports) {
 	loadedPromise.resolve();
 
 	var defaultSource;
-	var domainPromise = $.Deferred();
 	var domains = [];
 
-	sourceAPI.getSources().then(function (sources) {
-		if (sources.length == 0) {
-			sharedState.appInitializationStatus('no-sources-available');
-			return;
-		}
-		// find the source which has a Vocabulary Daimon with priority = 1
-		var prioritySources = sources.filter(function (source) {
-			return source.daimons.filter(function (daimon) {
-				return daimon.daimonType == "Vocabulary" && daimon.priority == "1"
-			}).length > 0
-		});
-		if (prioritySources.length > 0)
-			defaultSource = prioritySources[0];
-		else // find the first vocabulary or CDM daimon
-			defaultSource = sources.filter(function (source) {
-				return source.daimons.filter(function (daimon) {
-					return daimon.daimonType == "Vocabulary" || daimon.daimonType == "CDM"
-				}).length > 0
-			})[0];
-
-		// preload domain list once for all future calls to getDomains()
-		$.ajax({
-			url: config.webAPIRoot + 'vocabulary/' + defaultSource.sourceKey + '/domains',
-		}).then(function (results) {
-			$.each(results, function (i, v) {
-				domains.push(v.DOMAIN_ID);
-			});
-			domainPromise.resolve(domains);
+	getDomains();
```

Comment:
Why do you need to call it?

Reply:
This triggers the domain initialization when the module is first loaded so that it's 'ready' when we get to a place in the UI that wants the domain list (but don't want to make the users wait).

I found the other PR for 2.4's branch (or was it 2.5?) and I'd like to push some changes to this PR's branch to try to account for the origional design goal: to load the domain list once per application start (once per module load, which modules are only loaded once per app launch).  I'll push a commit to this PR's branch as a suggestion, and if you guys like, take it, if not, i'll revert it.

Stand by, should only take me a few minutes.

Reply:
@chrisknoll ,

1. You open Atlas app first time after deployment, there are no sources in the system
1. Your getDomains() function is automatically called, resulting in empty domains array
1. You go to Configuration tab, add Data source
1. Go to Cohort Definition / other place which relies on the domains list and observe incorrect behavior. Why is that a problem to load things on-demand and cache them. But on-demand, not on app startup?

Reply:
Yes, I pointed this out as a limitation:  if it only checks once per app load, then if you add sources later, you don't see the domains refresh (without further work to clear caches and whatnot).  The way to solve this would be to reset the domainPromise in the vocabAPI when sources get upadted.  But it just seems a lot of work for a 2.x series which will be probably quite different in 3.x, and the cost of reloading the app seems preferable to constantly polling the services + querying the vocabulary for the domains every time you want to get the domains (it's not a single service call it's 2: 1 to refresh the sources, and 2 to get the list of domains). Just doesn't feel worth it.

Reply:
Ok, so I got an idea, I'll push it up into a new commit but the idea is:  while we don't have sources, resolve a domainsPromise to []. Once we have a source with vocab, we can cache the domainPromise that resolves to the list of domains.  Shoudln't take long, stand by.